docs/production/authentication-methods.md

```diff
@@ -1069,6 +1069,67 @@ configure the JWT secret and algorithm via `JWT_AUTH_KEYS` in
 `/etc/zulip/settings.py`; see the inline comment documentation in that
 file for details.
 
+## Configuring a custom Python wrapper around the `authenticate` mechanism
+
+You can configure a custom function that will work as a wrapper around
+the `authenticate()` method calls of your enabled authentication backend
+classes. Knowledge of [how authentication backends work in Django][django-authenticate-details]
+as well as some familiarity with Zulip's authentication implementation in
+`zproject/backends.py` is required. When in doubt, reach out to us to ask
+for assistance or review of your custom authentication wrapper, especially
+considering that this is a security-sensitive feature and getting it wrong
+may be disastrous to your security model.
+
+In short, when a user is being authenticated to obtain a session
+(and thus this does NOT apply to authentication of API requests
+with a user's API key which use a separate authentication mechanism),
+the `authenticate()` methods of the various authentication backend classes
+in `zproject/backends.py` are called.
+
+You can write custom Python logic that will wrap such `authenticate()`
+calls by specifying in `/etc/zulip/settings.py`:
+
+```
+def custom_auth_wrapper(
+    auth_func, *args, **kwargs
+):
+    from zerver.lib.exceptions import JsonableError
+
+    backend = args[0]
+    backend_name = backend.name
+    request = args[1]
+    ip_address = request.META["REMOTE_ADDR"]
+    backend.logger.info("%s backend. ip is %s", backend_name, ip_address)
+
+    user_profile = auth_func(*args, **kwargs)
+    if user_profile is not None and user_profile.delivery_email == "protecteduser@example.com":
+        if backend_name == "email" and ip_address != "x.x.x.x":
+            raise JsonableError("Your IP address is not allowed to log in as this user.")
+
+    return user_profile
+
+# We need to actually specify to use this function defined above as the
+# custom authentication wrapper:
+CUSTOM_AUTHENTICATION_WRAPPER_FUNCTION = custom_auth_wrapper
+```
+
+`auth_func` is the underlying `authenticate` function belonging to the authentication
+backend class currently being processed. Therefore, your `custom_auth_wrapper` can inspect
```

Comment:
I think we need to be explicit that if you have multiple backends enabled, that means this function will be called more than once per login?

Reply:
Yeah that sounds right.

Reply:
Added this sentence here:

> If you have more than one backend enabled, this will be executed multiple times - each time with `auth_func` being the `authenticate` function of the backend class currently being processed.